crates/lib/src/qvm/mod.rs

```diff
@@ -32,73 +41,125 @@ impl QvmResultData {
+    #[cfg(feature = "tracing")]
+    tracing::debug!("parsing a program to be executed on the qvm");
+    let program = Program::from_str(quil).map_err(Error::Parsing)?;
+    run_program(
+        &program,
+        shots,
+        addresses,
+        params,
+        measurement_noise,
+        gate_noise,
+        rng_seed,
+        config,
+    )
+    .await
 }
 
-#[derive(Serialize, Debug, Clone, PartialEq, Eq)]
-#[serde(rename_all = "kebab-case")]
-struct Request<'request> {
-    quil_instructions: String,
-    addresses: HashMap<&'request str, bool>,
-    trials: u16,
-    #[serde(rename = "type")]
-    request_type: RequestType,
-}
+/// Run a [`Program`] on the QVM. The given [`Parameters`] are used to parametrize the value of
+/// memory locations across shots.
+#[allow(clippy::too_many_arguments)]
+pub async fn run_program(
+    program: &Program,
+    shots: u16,
+    addresses: HashMap<String, AddressRequest>,
+    params: &Parameters,
+    measurement_noise: Option<(f64, f64, f64)>,
+    gate_noise: Option<(f64, f64, f64)>,
+    rng_seed: Option<i64>,
+    config: &ClientConfiguration,
+) -> Result<QvmResultData, Error> {
+    #[cfg(feature = "tracing")]
+    tracing::debug!(
+        %shots,
+        ?addresses,
+        ?params,
+        "executing program on QVM"
+    );
+    if shots == 0 {
+        return Err(Error::ShotsMustBePositive);
+    }
 
-impl<'request> Request<'request> {
-    fn new(program: &str, shots: u16, readouts: &'request [Cow<'request, str>]) -> Self {
-        let addresses: HashMap<&str, bool> = readouts.iter().map(|v| (v.as_ref(), true)).collect();
-        Self {
-            quil_instructions: program.to_string(),
-            addresses,
-            trials: shots,
-            request_type: RequestType::Multishot,
+    // Create a clone of the program so MOVE statements can be prepended to it
+    let mut program = program.clone();
+
+    for (name, values) in params {
+        match program.memory_regions.get(name.as_ref()) {
+            Some(region) => {
+                if region.size.length != values.len() as u64 {
+                    return Err(Error::RegionSizeMismatch {
+                        name: name.clone(),
+                        declared: region.size.length,
+                        parameters: values.len(),
+                    });
+                }
+            }
+            None => {
+                return Err(Error::RegionNotFound { name: name.clone() });
+            }
+        }
+        for (index, value) in values.iter().enumerate() {
+            program.instructions.insert(
+                0,
+                Instruction::Move(Move {
+                    destination: ArithmeticOperand::MemoryReference(MemoryReference {
+                        name: name.to_string(),
+                        index: index as u64,
+                    }),
+                    source: ArithmeticOperand::LiteralReal(*value),
+                }),
+            );
         }
     }
```

Comment:
Might be missing context here, but this looks like prepending instructions into a `Vec`. That's going to involve shifting all of the existing instructions every time.

The outer loop could probably also be rewritten in terms of iterators. The benefit would be checking all regions _before_ modifying the cloned `Program` and essentially short-circuiting. Something like:

(This might break upon applying, because I originally wrote the suggestion based on just the added lines, which had mismatched braces)

```suggestion
    program.instructions = params.into_iter().map(|(name, values)| {
        match program.memory_regions.get(name.as_ref()) {
            Some(region) => {
                if region.size.length == values.len() as u64 {
                    Ok((name, values))
                } else {
                    Err(Error::RegionSizeMismatch {
                        name,
                        declared: region.size.length,
                        parameters: values.len(),
                    })
                }
            }
            None => {
                Err(Error::RegionNotFound { name: name.clone() });
            }
        }
    }).collect::<Result<Vec<_>, _>>()?
        .into_iter()
        .rev()
        .flat_map(|(name, values)| {
            values.iter().enumerate().rev().map(|(index, value)| {
                Instruction::Move(Move {
                    destination: ArithmeticOperand::MemoryReference(MemoryReference {
                        name: name.to_string(),
                        index: index as u64,
                    }),
                    source: ArithmeticOperand::LiteralReal(*value),
                })
            })
        }).chain(program.instructions).collect();
```

- `rev()` because, with the current logic, the last item ends up first in the list. To keep the same ordering, we need to reverse both the outer and inner iterators
- We should only need to allocate memory once (per `collect`), since iterators allow providing size hints, and iterators over slices should provide those hints, so internally it should be able to use `Vec::with_capacity`.

Reply:
Okay, the refactor is done. Ended up with a similar approach, but opted to use a `try_for_each` to check the precondition since it felt more clear and was one less `collect`. Also, didn't use `rev()` since the ordering of the `MOVE` statements doesn't matter.